rep-2007.rst

```diff
@@ -0,0 +1,246 @@
+Whereas without this feature, even to send a ``cv::Mat`` from one publisher to a subscription in the same process would require converting it to a ``sensor_msgs::msg::Image`` first and then back again to cv::Mat.
+
+
+Terminology
+===========
+
+.. glossary::
+     Custom type (in code, ``CustomType``)
+       A data structure that *is not* a ROS 2 interface, such as ``std::string`` or ``cv::Mat``.
+     ROS type (in code, ``RosType``)
+       A data structure that *is* a ROS 2 interface, such as ``std_msgs::msg::String`` or ``sensor_msgs::msg::Image``.
+
+
+Specification
+=============
+
+There are different tiers of expressivity in the syntax for type adaptation, including both explicit to implicit forms.
+The explicit forms require the custom type and the ROS type to be specified; 
+while the implicit form makes an assumption about the ROS type based on the custom type.
+The explicit forms may be used to clearly communicate what is happening, which could be useful in examples or tutorials, or may be used depending on the preferences of the developer.
+While the implicit, and most concise, form will primarily be used for its convenience (although it lacks some of the functionality of the explicit forms).
+
+Below are examples of different syntaxes for type adaptation.
+In these examples and using the terminology above, ``std::string`` will be considered the custom type and the ROS interface ``std_msgs::msg::String`` will be considered the ROS type.
+Note that, for every publisher created in the examples below, a ``std::string`` type variable (the custom type) would be adapted to be published on the ROS network as a ``std_msgs::msg::String`` (the ROS type).
+
+.. code-block:: cpp
+
+   // Explicit Form 1: Most explicit
+   using MsgT = rclcpp::AdaptType<std::string>::to<std_msgs::msg::String>;
+   auto pub1 = std::make_shared<rclcpp::Publisher<MsgT>>();
+
+The most explicit form could be shortened in the following ways:
+
+.. code-block:: cpp
+
+   // Explicit Form 2
+   using MsgT = rclcpp::AdaptType<std::string, std_msgs::msg::String>;
+   auto pub2 = std::make_shared<rclcpp::Publisher<MsgT>>();
+
+   // Explicit Form 3
+   auto pub3 = std::make_shared<rclcpp::Publisher<std::string, std_msgs::msg::String>>();
+
+Note that the 3rd explicit form may be prohibitively complicated to implement since there are multiple template arguments already for publishers and subscribers.
+
+The final shorthand performs an implicit type adaptation with the custom type ``std::string`` which will be adapted to the ROS type``std_msgs::msg::String``:
+
+.. code-block:: cpp
+
+   // Implicit Form
+   auto pub4 = std::make_shared<rclcpp::Publisher<std::string>>();
+
+In this final form, if the custom type (``std::string``) has multiple ROS types that it could represent, then an error will be raised during compilation.
```

Comment:
I was going to comment this.
I slightly prefer not having an implicit adapter, though I don't mind strongly.

Reply:
My current thinking is that supporting one custom type to many ROS types will not be possible, as it would require some kind of side effect in the template meta programming, which isn't possible.

So, though implicit, it will not be possible to include two "TypeAdaptors" for the same custom type in the first place. This (draft) REP hasn't been updated to reflect this.

Reply:
> My current thinking is that supporting one custom type to many ROS types will not be possible, as it would require some kind of side effect in the template meta programming

I don't completely understand the limitation, but I would rather have an explicit and more verbose "TypeAdaptor" than not being able to support two "TypeAdaptor"s for the same type (if that's the trade-off).